### PR TITLE
[SPARK-26132][BUILD][CORE] Remove support for Scala 2.11 in Spark 3.0.0

### PR DESCRIPTION
**Pull request by srowen**
## What changes were proposed in this pull request?

Remove Scala 2.11 support in build files and docs, and in various parts of code that accommodated 2.11. See some targeted comments below.

## How was this patch tested?

Existing tests.

### CODE REVIEW

File: R/pkg/R/sparkR.R
Comment (srowen) [inline R/pkg/R/sparkR.R]: @felixcheung is it OK to refer to _2.12 artifacts here? I don't think this one actually exists, but is it just an example?

File: bin/load-spark-env.cmd
Comment (srowen) [inline bin/load-spark-env.cmd]: @gengliangwang this was the update I was talking about to the .cmd script. You can follow up with this change, uncommented, if you like, separately from this PR.

File: dev/create-release/release-build.sh
Comment (srowen) [inline dev/create-release/release-build.sh]: @vanzin @cloud-fan you may want to look at this. It's getting a little hairy in this script.

I recall that the goal was to use this script to create older Spark releases, so it needs logic for older versions. But looking at it, I don't think it actually creates quite the same release as older versions anyway. Is it OK to clean house here and assume only Spark 3 will be built from this script? I already deleted some really old logic here (Spark < 2.2)

File: project/MimaExcludes.scala
Comment (srowen) [inline project/MimaExcludes.scala]: I figure we can delete old Spark 2.2 and earlier excludes, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: @maryannxue I think you helped work on this part ... if we're only on Scala 2.12 now can we simplify this further?

File: bin/load-spark-env.cmd
Comment (gengliangwang) [inline bin/load-spark-env.cmd]: I am not familiar with .cmd script. Should we keep the quote here, `"2.12"`

Comment (HyukjinKwon) [reply bin/load-spark-env.cmd]: Nope, it takes as is including quotes ... haha odd (to me).

Comment (srowen) [reply bin/load-spark-env.cmd]: Oh I see, so we shouldn't add quotes to values like SPARK_ENV_CMD above, but use them in if conditions, call, etc?

Comment (HyukjinKwon) [reply bin/load-spark-env.cmd]: Here, I ran some of simple commands:

```cmd
C:\>set A=aa

C:\>ECHO %A%
aa

C:\>set A="aa"

C:\>ECHO %A%
"aa"

C:\>call "python.exe"
Python 3.6.4 (v3.6.4:d48eceb, Dec 19 2017, 06:54:40) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> exit(0)

C:\>call python.exe
Python 3.6.4 (v3.6.4:d48eceb, Dec 19 2017, 06:54:40) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> exit(0)
```

File: R/pkg/R/sparkR.R
Comment (vanzin) [reply R/pkg/R/sparkR.R]: I think there's even a separate discussion about even using this as an example, since that package is now in Spark since 2.4.

Comment (srowen) [reply R/pkg/R/sparkR.R]: @felixcheung was the conclusion that we can make this a dummy package? I just want to avoid showing _2.11 usage here.

File: dev/create-release/release-build.sh
Comment (srowen) [reply dev/create-release/release-build.sh]: Just a question though, if you're releasing 2.3 wouldn't you use the release script as of 2.3? 

I think the script works for Spark 1.6 and 2.x, but not earlier versions, already. What about drawing the lines at major releases, so that this can simplify further?

Right now I think it's basically trying to support 2.2+, which are the non-EOL releases, which seems reasonable?

Comment (vanzin) [reply dev/create-release/release-build.sh]: >  if you're releasing 2.3 wouldn't you use the release script as of 2.3?

I started like that when I RM'ed 2.3.1, but the scripts were kinda broken that I had to fix a bunch of stuff in master. At that point, it didn't make sense to use the scripts in the 2.3 branch anymore, and keeping them in sync was kinda wasted effort.

> basically trying to support 2.2+, which seems reasonable?

That's my view. The release scripts in master should be usable for all non-EOL releases.

File: R/pkg/R/sparkR.R
Comment (felixcheung) [reply R/pkg/R/sparkR.R]: yes, dummy name is completely fine with me.

**Comment by srowen** [comment]
Note I'm holding on to this PR for a while as I understand it might be disruptive to downstream builds to remove 2.11 support just now. Will look at merging it in weeks. Right now it's an FYI.

**Comment by srowen** [comment]
Still holding on to this. I intend to merge it before Spark 3, but have gotten feedback that it would be helpful to hold off for a while.

**Review by dbtsai — APPROVED** [review]
LGTM for removing Scala 2.11.

**Comment by srowen** [comment]
Rebased to keep this up to date; not merging just yet

**Comment by mengxr** [comment]
@srowen @dbtsai Have we decided to remove Scala 2.11 support in 3.0? If not, shall we start a vote?

**Comment by srowen** [comment]
I think there is support for it, and don't know if we need a vote. Part of the reason I delay is that it's not a hurry, and in a way I wouldn't mind keeping Spark 3 pretty nearly compatible with 2.11, even if we won't support it. Leaving the 2.11 build maybe lets us avoid unnecessary incompatibilities with 2.11. I was planning to check on support again for this in a month or so.

**Comment by skonto** [comment]
@srowen +100 for removing 2.11, let's move forward, 2.11 is old and Spark 3.0.0 gives us an opportunity to get rid of it.

**Comment by dbtsai** [comment]
+1

**Comment by srowen** [comment]
Going to merge this if the next set of test passes. Last call for comments.

**Comment by srowen** [comment]
Merged to master

**Comment by jzhuge** [comment]
@srowen So awesome to merge this. Does it mean we can revert `[SPARK-27250][TEST-MAVEN][BUILD] Scala 2.11 maven compile should target Java 1.8` now?

**Comment by srowen** [comment]
@jzhuge no that change is still valid. We still need to target Java 1.8 now, and will need to continue to do so even when Java 11 support works

**Comment by kiszk** [comment]
@sowen I realized that this PR causes an error when we execute `do-release-docker.sh` for branch-2.3.
```
...
Checked out revision 35358.
Copying release tarballs
cp: cannot stat 'pyspark-*': No such file or directory
```

For Spark 2.4, `PUBLISH_SCALA_2_11 = 1, PUBLISH_SCALA_2_10=1`   
For Spark 2.3, `PUBLISH_SCALA_2_11 = 1, PUBLISH_SCALA_2_10=0`
Since [This change](https://github.com/apache/spark/commit/8bc304f97ee693b57f33fa6708eb63e2d641c609#diff-01ca42240614718522afde4d4885b40dR293) does not pass `"withpip,withr"`, the script does not build `pyspark-*` and `SparkR_*`.

**Comment by dongjoon-hyun** [comment]
Hi, @kiszk . You should not use `master` branch release script.
For `branch-2.4`, I'm using the release script in `branch-2.4` as you see my VOTE email.

cc @gatorsmile , @vanzin

**Comment by dongjoon-hyun** [comment]
BTW, that's the reason why we made `2.4.3` urgently after `2.4.2`.

**Comment by dongjoon-hyun** [comment]
FYI, I'm blocked by recent Jekyll release issue. I'll make a PR soon against `master` and `branch-2.4`. I'll ping you on there, too.

**Comment by kiszk** [comment]
@dongjoon-hyun Thank you for your suggestion. Got it for `2.4.3`.
I will use the script in `branch-2.3`.

**Comment by kiszk** [comment]
I will add it to `https://spark.apache.org/release-process.html`

**Comment by dongjoon-hyun** [comment]
BTW, I didn't test `branch-2.3` release script before. As you know, I've been only testing on `branch-2.4`. Please test it before releasing.

**Comment by kiszk** [comment]
[`Branch-2.3`](https://github.com/apache/spark/tree/branch-2.3/dev/create-release) is pretty old. There is no script `do-release-docker.sh`. I will check the old page.

**Comment by dongjoon-hyun** [comment]
Although this is Apache Spark 2.3.x EOL release, you need to commit to `branch-2.3` whatever you do. Otherwise, it will not be reproducible for the others.

**Comment by kiszk** [comment]
Yes, you are right. If backporting of `do-release-docker.sh` works well, I will commit it to `branch-2.3`.

**Pull request closed without merging**
